crates/pallets/starknet/src/lib.rs

```diff
@@ -1106,6 +1106,28 @@ impl<T: Config> Pallet<T> {
         Ok(results)
     }
 
+    pub fn estimate_message_fee(message: HandleL1MessageTransaction) -> Result<(u64, u64), DispatchError> {
+        let chain_id = Self::chain_id();
+        let transaction = message.into_executable::<T::SystemHash>(chain_id, Fee::default(), true);
+
+        let tx_execution_infos = transaction
+            .execute(
+                &mut BlockifierStateAdapter::<T>::default(),
+                &Self::get_block_context(),
+                &RuntimeExecutionConfigBuilder::new::<T>().with_query_mode().build(),
+            )
+            .map_err(|e| {
+                log::error!("failed to execute invoke tx: {:?}", e);
+                Error::<T>::TransactionExecutionFailed
+            })?;
+
+        if let Some(l1_gas_usage) = tx_execution_infos.actual_resources.0.get("l1_gas_usage") {
+            Ok((tx_execution_infos.actual_fee.0 as u64, *l1_gas_usage))
+        } else {
+            Err(Error::<T>::TransactionExecutionFailed.into())
```

Comment:
If you rebase you will find a dedicated error message for l1_gas_usage field missing

Reply:
Oh no sorry. It's a PR of mine which is not merged yet.